src/structstore_shared.hpp

```
@@ -42,62 +44,192 @@ class StructStoreShared {
-            }
-            result = read(shm_fd, &size, sizeof(size_t));
-            if (result != sizeof(size_t)) {
-                throw std::runtime_error("reading original size failed");
-            }
-            if (size < sizeof(SharedData)) {
-                throw std::runtime_error("original size is invalid");
-            }
-            lseek(shm_fd, 0, SEEK_SET);
-            shm_ptr = (SharedData*) mmap(original_ptr, size, PROT_READ | PROT_WRITE,
-                                         MAP_SHARED | MAP_FIXED_NOREPLACE, shm_fd, 0);
-            if (shm_ptr == MAP_FAILED || shm_ptr != original_ptr) {
-                throw std::runtime_error("mmap'ing existing memory failed");
-            }
-            if (shm_ptr->original_ptr != original_ptr) {
-                throw std::runtime_error("inconsistency detected");
+
+        if (reinit && shm_stat.st_size != 0) {
+            // we found an opened memory segment with a non-zero size,
+            // it's likely an old segment thus ...
+
+            // ... we open it and mark it as closed ...
+            mmap_existing_shm();
+            shm_ptr->ok.store(false);
+            shm_ptr->usage_count -= 1;
+
+            // ... then unmap it, ...
+            munmap(shm_ptr, shm_ptr->size);
+            shm_ptr = nullptr;
+
+            // ... then unlink it, ...
+            shm_unlink(shm_path.c_str());
+            close(shm_fd);
+
+            // ... and finally recreate it
+            shm_fd = shm_open(shm_path.c_str(), O_CREAT | O_RDWR, 0600);
+
+            if (-1 == shm_fd) {
+                throw std::runtime_error("opening shared memory failed");
             }
-            ++shm_ptr->usage_count;
-        } else {
-            std::cout << "reserving new memory ..." << std::endl;
+        } else if (!created && shm_stat.st_mode != 0100660) {
+            // shared memory is not ready for opening yet
+            return;
```

This leaves the `StructStoreShared` object in an inconsistent state (`shm_ptr == nullptr`) and should throw an error instead